### ISSUE TITLE
How does the sequence of conditions in 'SQL where' affect the query result?  

### ISSUE THREAD

**Issue by titawork** [heavyai/heavydb#373]
I execute SQL as following,
****1**, running well**
select ti_c_csdcszsjsmx1.mxcjrq  from
ti_c_csdcszsjsmx1,
tf_c_tradeseat
where
ti_c_csdcszsjsmx1.fund_id = tf_c_tradeseat.fund_id and
ti_c_csdcszsjsmx1.mxjydy = tf_c_tradeseat.seat_code and
tf_c_tradeseat.market_no = 2;

****2**. running well**
select ti_c_csdcszsjsmx1.mxcjrq  from
ti_c_csdcszsjsmx1,
tf_c_tradeseat
where
tf_c_tradeseat.market_no = 2 and
ti_c_csdcszsjsmx1.fund_id = tf_c_tradeseat.fund_id;

**3. Error** 
select ti_c_csdcszsjsmx1.mxcjrq  from
ti_c_csdcszsjsmx1,
tf_c_tradeseat
where
tf_c_tradeseat.market_no = 2 and
ti_c_csdcszsjsmx1.fund_id = tf_c_tradeseat.fund_id and
ti_c_csdcszsjsmx1.mxjydy = tf_c_tradeseat.seat_code;

it is odd.

**server version: 4.7.1
client version:4.7.1**

Tables as following,
tf_c_tradeseat:
fund_id BIGINT,
market_no INTEGER ENCODING FIXED(16),

 ti_c_csdcszsjsmx1:
fund_id BIGINT, 
mxjydy TEXT ENCODING DICT(32)

**logs when error:**
I0812 09:09:13.372398 13465 Calcite.cpp:432] User clear catalog clear_db sql 'select ti_c_csdcszsjsmx1.mxcjrq  from ti_c_csdcszsjsmx1, tf_c_tradeseat where tf_c_tradeseat.market_no = 2 and ti_c_csdcszsjsmx1.fund_id = tf_c_tradeseat.fund_id and ti_c_csdcszsjsmx1.mxjydy = tf_c_tradeseat.seat_code;'
I0812 09:09:13.376538 13465 Calcite.cpp:452] Time in Thrift 0 (ms), Time in Java Calcite server 3 (ms)
I0812 09:09:13.377110 13465 RelAlgExecutor.cpp:640] Assigned input (RelScan<139644050853968> ti_c_csdcszsjsmx1) to nest level 0
I0812 09:09:13.377125 13465 RelAlgExecutor.cpp:640] Assigned input (RelScan<139644042936608> tf_c_tradeseat) to nest level 1
I0812 09:09:13.377140 13465 RelAlgExecutor.cpp:640] Assigned input (RelScan<139644050853968> ti_c_csdcszsjsmx1) to nest level 0
I0812 09:09:13.377142 13465 RelAlgExecutor.cpp:640] Assigned input (RelScan<139644042936608> tf_c_tradeseat) to nest level 1
E0812 09:09:13.381548 13465 MapDHandler.cpp:4958] Exception: Cast from dictionary-encoded string to none-encoded would be slow
I0812 09:09:13.381680 13465 MapDHandler.cpp:5620] stdlog MapDHandler.cpp 748 sql_execute 9 clear_db clear 202-8pIt {"query_str"} {"select ti_c_csdcszsjsmx1.mxcjrq  from ti_c_csdcszsjsmx1, tf_c_tradeseat where tf_c_tradeseat.market_no = 2 and ti_c_csdcszsjsmx1.fund_id = tf_c_tradeseat.fund_id and ti_c_csdcszsjsmx1.mxjydy = tf_c_tradeseat.seat_code;"}

**logs when right:**
I0812 09:11:05.481544 13465 Calcite.cpp:432] User clear catalog clear_db sql 'select ti_c_csdcszsjsmx1.mxcjrq  from ti_c_csdcszsjsmx1, tf_c_tradeseat where ti_c_csdcszsjsmx1.fund_id = tf_c_tradeseat.fund_id and ti_c_csdcszsjsmx1.mxjydy = tf_c_tradeseat.seat_code and tf_c_tradeseat.market_no = 2;'
I0812 09:11:05.486824 13465 Calcite.cpp:452] Time in Thrift 1 (ms), Time in Java Calcite server 3 (ms)
I0812 09:11:05.487359 13465 RelAlgExecutor.cpp:640] Assigned input (RelScan<139644048504448> ti_c_csdcszsjsmx1) to nest level 0
I0812 09:11:05.487371 13465 RelAlgExecutor.cpp:640] Assigned input (RelScan<139644027069216> tf_c_tradeseat) to nest level 1
I0812 09:11:05.487385 13465 RelAlgExecutor.cpp:640] Assigned input (RelScan<139644048504448> ti_c_csdcszsjsmx1) to nest level 0
I0812 09:11:05.487390 13465 RelAlgExecutor.cpp:640] Assigned input (RelScan<139644027069216> tf_c_tradeseat) to nest level 1
I0812 09:11:05.493001 13465 MapDHandler.cpp:5620] stdlog MapDHandler.cpp 748 sql_execute 11 clear_db clear 202-8pIt {"query_str","execution_time_ms","total_time_ms"} {"select ti_c_csdcszsjsmx1.mxcjrq  from ti_c_csdcszsjsmx1, tf_c_tradeseat where ti_c_csdcszsjsmx1.fund_id = tf_c_tradeseat.fund_id and ti_c_csdcszsjsmx1.mxjydy = tf_c_tradeseat.seat_code and tf_c_tradeseat.market_no = 2;","10","11"}

**Comment by cdessanti** [comment]
The seatcode datatype is...? You forgot on ddl.
If it a dict encoded try to share the dictionary between the two columns you are joining to resolve the issue

Ottieni Outlook per Android<https://aka.ms/ghei36>

________________________________

**Comment by titawork** [comment]
Yes. 
 tf_c_tradeseat.seat_code TEXT ENCODING DICT(32) the same as ti_c_csdcszsjsmx1.mxjydy
Could you share me the root cause？
And Should I share the dictionary for all of other columns like this two columns?

**Comment by cdessanti** [comment]
I noticed that sometimes to get rid of this kind of errors I had to share the di tory between columns I was comparing. It's weird so I guess it's a bug or something that has to be reworked, but expecially if cardinality is high you should be also a performance boost sharing the dictionary

Ottieni Outlook per Android<https://aka.ms/ghei36>

________________________________

**Comment by titawork** [comment]
Yes. I also think it is a bug even shared dictionary can be a workaround.

**Comment by titawork** [comment]
Do Omnisci have plan to fix this issue? 
I think everyone can execute such SQL but without shared dictionary. They will also run into this problem.

**Comment by cdessanti** [comment]
Have you tried sharing the dictionary of the keys?
It is a sort of best practice

Ottieni Outlook per Android<https://aka.ms/ghei36>

________________________________

**Comment by titawork** [comment]
Yes. I tried and it works.
Okey， it is better than nothing -_-

**Comment by alexbaden** [comment]
Hi @titawork 

The error you are seeing is a watchdog exception -- you can choose to conditionally disable the watchdog at runtime by passing `enable-watchdog = false` in your config file. But, typically those exceptions are a warning that something you are about to do will be very slow -- and indeed, if I recreate your tables and run with watchdog disabled I see that, for some reason, we are not using the hash join framework for the string comparison and are indeed converting all the dictionary encoded strings to none-encoded strings, putting them in a buffer, and comparing them. 

We should be able to accelerate this query and remove the error message by resolving that behavior -- since your first query works, I think we likely need to have some sort of ordering or preference when evaluating which quals we are going to build a hash table over. We'll take a look and push something out soon. Thanks for the bug report!

**Comment by titawork** [comment]
Hi @alexbaden
Thank you very much for your so detail response.
Looking forward the updating coming soon.

**Comment by titawork** [comment]
Hi @alexbaden
Please let me ask one more question, In my work based on omnisci, I have a lot of string comparison and group by with string . Normally the quantity of rows can be 10 million.  Should I use TEXT ENCODING NONE or TEXT ENCODING DICT ? Which one is better for my situation? Thank you very much!

**Comment by randyzwitch** [comment]
@titawork Unless your string data are unique across nearly every single row (similar to a primary key or other non-repeating value), using TEXT ENCODING DICT is usually better performing and more appropriate

**Comment by titawork** [comment]
Understood！Thank you very much !

**Comment by psilvest** [comment]
This issue has been inactive for a while, and the project has changed substantially. If this still occurs on the latest release, please add reproduction steps and environment details.
